CarpetX/src/valid.cxx

```diff
@@ -477,33 +481,50 @@ void check_valid_ga(const int gi, const int vi, const int tl,
 
   auto &restrict globaldata = ghext->globaldata;
   auto &restrict arraygroupdata = *globaldata.arraygroupdata.at(gi);
+  cGroup group;
+  int ierr = CCTK_GroupData(gi, &group);
+  assert(!ierr);
 
   std::size_t nan_count{0};
 
   const valid_t &valid = arraygroupdata.valid.at(tl).at(vi).get();
   if (valid.valid_all())
     return;
 
-  CCTK_REAL poison;
-  std::memcpy(&poison, &ipoison, sizeof poison);
+  const size_t typesize = size_t(CCTK_VarTypeSize(group.vartype));
+  char poison[typesize];
+  // for int: deadbeef for little endian machines
+  for (size_t i = 0; i < typesize; i += sizeof poison)
+    std::memcpy(&poison[i], &ipoison, std::min(sizeof poison, typesize - i));
 
   // arrays have no boundary so we expect them to alway be valid
   assert(valid.valid_outer && valid.valid_ghosts);
 
   if (valid.valid_int) {
-    const CCTK_REAL *restrict const ptr = &arraygroupdata.data.at(tl).at(vi);
     int dimension = arraygroupdata.dimension;
     const int *gsh = arraygroupdata.gsh;
     int n_elems = 1;
     for (int i = 0; i < dimension; i++)
       n_elems *= gsh[i];
-    for (int i = 0; i < n_elems; i++) {
-      using std::isnan;
-      if (CCTK_BUILTIN_EXPECT(nan_handling == nan_handling_t::allow_nans
-                                  ? std::memcmp(ptr, &poison, sizeof *ptr) == 0
-                                  : isnan(*ptr),
-                              false))
-        ++nan_count;
+    if (group.vartype == CCTK_VARIABLE_REAL) {
+      const CCTK_REAL *restrict const ptr = static_cast<const CCTK_REAL *const>(
+          arraygroupdata.data.at(tl).data_at(vi));
+      for (int i = 0; i < n_elems; i++) {
+        using std::isnan;
+        if (CCTK_BUILTIN_EXPECT(nan_handling == nan_handling_t::allow_nans
+                                    ? std::memcmp(ptr, poison, sizeof *ptr) == 0
+                                    : isnan(*ptr),
```

Comment:
You could call `isnan(real) || isnan(imag)` for complex numbers.

Reply:
I guess, yes.

Reply:
done in 49de23b